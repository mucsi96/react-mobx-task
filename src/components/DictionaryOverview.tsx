import * as React from 'react';
import { inject, observer } from 'mobx-react';
import DictionaryStore from '../stores/DictionaryStore';
import ViewStore from '../stores/ViewStore';
import Dictionary from '../models/Dictionary';
import { transaction } from 'mobx';

interface DictionaryProps {
  dictionaryId: string;
}

@inject('dictionaryStore', 'viewStore')
@observer
export default class DictionaryOverview extends React.Component<
  DictionaryProps
> {
  get injected() {
    return (this.props as unknown) as {
      dictionaryStore: DictionaryStore;
      viewStore: ViewStore;
    };
  }

  private readonly dictionary: Dictionary;

  constructor(props: DictionaryProps) {
    super(props);
    this.dictionary = this.injected.dictionaryStore.getDictionaryById(
      props.dictionaryId
    );
  }

  public render(): JSX.Element {
    if (!this.dictionary.transformations.length) {
      return (
        <>
          {this.renderHeader()}
          <div className="ui placeholder segment">
            <div className="ui icon header">No transformations yet.</div>
            {this.renderAddButton()}
          </div>
        </>
      );
    }

    return (
      <>
        {this.renderHeader()}
        <div className="ui basic segment">
          <table className="ui selectable celled table">
            <thead>
              <tr>
                <th>From</th>
                <th>To</th>
              </tr>
            </thead>
            <tbody>
              {this.dictionary.transformations.map(({ id, from, to }) => (
                <tr key={id} onClick={() => this.handleItemClick(id)}>
                  <td>
                    {this.renderValidationIcon(id)}
                    {from}
                  </td>
                  <td>{to}</td>
                </tr>
              ))}
            </tbody>
          </table>
          {this.renderAddButton()}
        </div>
      </>
    );
  }

  private renderHeader() {
    return (
      <div className="ui basic clearing segment">
        <div className="ui left floated header">{this.dictionary.name}</div>
        <div className="ui right floated header">
          <button className="ui button" onClick={this.handleClickEdit}>
            Edit
          </button>
          <button className="ui red button" onClick={this.handleClickDelete}>
            Delete
          </button>
        </div>
      </div>
    );
  }

  private renderAddButton() {
    return (
      <button
        autoFocus={true}
        className="ui primary button"
        onClick={this.handleClickAdd}
      >
        Add transformation
      </button>
    );
  }

  private handleClickAdd = () => {
    this.injected.viewStore.createTransformation(this.props.dictionaryId);
  };

  private handleClickEdit = () => {
    this.injected.viewStore.editDictionary(this.props.dictionaryId);
  };

  private handleClickDelete = () => {
    this.injected.dictionaryStore.removeDectionary(this.props.dictionaryId);
    this.injected.viewStore.showOverview();
  };

  private handleItemClick = (transformationId: string) => {
    this.injected.viewStore.editTransformation(
      this.props.dictionaryId,
      transformationId
    );
  };

  private renderValidationIcon(transformationId: string) {
    const validationMessage = this.dictionary.getValidationErrorById(
      transformationId
    );

    if (validationMessage) {
      return (
        <span className="ui vaidation icon" data-tooltip={validationMessage}>
          <i className="yellow exclamation triangle icon" />
        </span>
      );
    }

    return null;
  }
}
