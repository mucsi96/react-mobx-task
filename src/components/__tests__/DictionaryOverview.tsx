import * as React from 'react';
import * as RTL from 'react-testing-library';
import { Provider } from 'mobx-react';
import DictionaryOverview from '../DictionaryOverview';

const testDictionary = {
  id: '1',
  name: 'a',
  transformations: [
    { id: '1', from: 'b', to: 'c' },
    { id: '2', from: 'd', to: 'e' }
  ],
  getValidationErrorById: (id: string) =>
    id === '1' ? undefined : 'test validation error'
};

const dictionaryStore = {
  removeDectionary: jest.fn(),
  getDictionaryById: (id: string) => (id === '7' ? testDictionary : undefined)
};

const viewStore = {
  createTransformation: jest.fn(),
  editDictionary: jest.fn(),
  showOverview: jest.fn(),
  editTransformation: jest.fn()
};

const render = () =>
  RTL.render(
    <Provider dictionaryStore={dictionaryStore} viewStore={viewStore}>
      <DictionaryOverview dictionaryId="7" />
    </Provider>
  );

describe('DictionaryList', () => {
  it('renders', () => {
    const { asFragment } = render();
    expect(asFragment()).toMatchSnapshot();
  });

  it('navigates to create transformation view on "Add transformation" click', () => {
    viewStore.createTransformation.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Add transformation'));
    expect(viewStore.createTransformation).toBeCalledWith('7');
  });

  it('navigates to edit dictionary view on "Edit" click', () => {
    viewStore.editDictionary.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Edit'));
    expect(viewStore.editDictionary).toBeCalledWith('7');
  });

  it('removes dictionary on "Delete" click', () => {
    dictionaryStore.removeDectionary.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Delete'));
    expect(dictionaryStore.removeDectionary).toBeCalledWith('7');
  });

  it('navigates to overview on "Delete" click', () => {
    viewStore.showOverview.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Delete'));
    expect(viewStore.showOverview).toBeCalled();
  });

  it('navigates to edit transformation view on item click', () => {
    viewStore.editTransformation.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('d'));
    expect(viewStore.editTransformation).toBeCalledWith('7', '2');
  });

  it('navigates to overview on "Back" click', () => {
    viewStore.showOverview.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Back'));
    expect(viewStore.showOverview).toBeCalled();
  });
});
