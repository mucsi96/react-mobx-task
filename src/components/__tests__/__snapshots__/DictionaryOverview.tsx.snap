// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DictionaryList renders 1`] = `
<DocumentFragment>
  <div
    class="ui basic clearing segment"
  >
    <div
      class="ui left floated header"
    >
      a
    </div>
    <div
      class="ui right floated header"
    >
      <button
        class="ui button"
      >
        Back
      </button>
      <button
        class="ui button"
      >
        Edit
      </button>
      <button
        class="ui red button"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    class="ui basic segment"
  >
    <table
      class="ui selectable celled table"
    >
      <thead>
        <tr>
          <th>
            From
          </th>
          <th>
            To
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            b
          </td>
          <td>
            c
          </td>
        </tr>
        <tr>
          <td>
            <span
              class="ui vaidation icon"
              data-tooltip="test validation error"
            >
              <i
                class="yellow exclamation triangle icon"
              />
            </span>
            d
          </td>
          <td>
            e
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="ui primary button"
    >
      Add transformation
    </button>
  </div>
</DocumentFragment>
`;
